src/modm/driver/encoder/encoder_slave_impl.hpp

```diff
@@ -0,0 +1,80 @@
+// coding: utf-8
+/*
+ * Copyright (c) 2019, Raphael Lehmann
+ *
+ * PositionTypehis file is part of the modm project.
+ *
+ * PositionTypehis Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef	ENCODER_SLAVE_HPP
+#	error	"Don't include this file directly, use 'encoder_slave.hpp' instead!"
+#endif
+#include "encoder_slave.hpp"
+
+template < class PinA, class PinB, typename PositionType, class Timeout, uint32_t period_ms >
+modm::EncoderSlave<PinA, PinB, PositionType, Timeout, period_ms>::EncoderSlave() :
+	setpoint(0), actualValue(0), timeout(0), state(State::State0)
+{
+	PinA::setOutput();
+	PinB::setOutput();
+	PinA::set();
+	PinB::set();
+	timeout.restart(period_ms);
+}
+
+template < class PinA, class PinB, typename PositionType, class Timeout, uint32_t period_ms >
+void
+modm::EncoderSlave<PinA, PinB, PositionType, Timeout, period_ms>::update()
+{
+	if(timeout.execute()){
+		if(setpoint > actualValue) {
+			// generate forward tick
+			switch(state) {
+				case State::State0:
+					PinA::reset();
+					state = State::State1;
+					break;
+				case State::State1:
+					PinB::reset();
+					state = State::State2;
+					break;
+				case State::State2:
+					PinA::set();
+					state = State::State3;
+					break;
+				case State::State3:
+					PinB::set();
+					state = State::State0;
+					break;
+			}
+			actualValue++;
+		}
+		else if(setpoint < actualValue) {
+			// generate backward tick
+			switch(state) {
+				case State::State0:
+					PinB::reset();
+					state = State::State3;
+					break;
+				case State::State1:
+					PinA::reset();
+					state = State::State0;
+					break;
+				case State::State2:
+					PinB::set();
+					state = State::State1;
+					break;
+				case State::State3:
+					PinA::set();
+					state = State::State2;
+					break;
+			}
+			actualValue--;
+		}
+		timeout.restart(period_ms);
```

Comment:
Why not use a `modm::PeriodicTimer` instead?

Reply:
Because it should be easier to replace the `modm::Timeout` with a custom class that uses e.g. a stm32 hardware timer as time source.

Reply:
That's not how it works, you implement a custom `modm::Clock` and pass it to `modm::Generic{Timeout, PeriodicTimer}`, so the amount of work is the same.